File: api/sync.js
const FILENAME = "lighting-schedule-data.json";

module.exports = async (req, res) => {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");

  if (req.method === "OPTIONS") return res.status(200).end();

  const token = process.env.BLOB_READ_WRITE_TOKEN;
  if (!token) return res.status(500).json({ error: "BLOB_READ_WRITE_TOKEN not set" });

  try {
    if (req.method === "GET") {
      const listRes = await fetch(`https://blob.vercel-storage.com?prefix=${FILENAME}`, {
        headers: { "Authorization": `Bearer ${token}` }
      });
      const listData = await listRes.json();
      if (!listData.blobs || listData.blobs.length === 0) return res.status(200).json({});
      const dataRes = await fetch(listData.blobs[0].url);
      if (!dataRes.ok) return res.status(200).json({});
      const data = await dataRes.json();
      return res.status(200).json(data);
    }

    if (req.method === "POST") {
      const incoming = req.body;

      // Load existing data first
      let existing = {};
      try {
        const listRes = await fetch(`https://blob.vercel-storage.com?prefix=${FILENAME}`, {
          headers: { "Authorization": `Bearer ${token}` }
        });
        const listData = await listRes.json();
        if (listData.blobs && listData.blobs.length > 0) {
          const dataRes = await fetch(listData.blobs[0].url);
          if (dataRes.ok) existing = await dataRes.json();
        }
      } catch(e) {}

      // Merge projects by ID - each project is identified by its id
      const existingProjects = existing.projects || [];
      const incomingProjects = incoming.projects || [];

      // Merge projects by ID - newer updatedAt wins
      const projectMap = {};
      existingProjects.forEach(p => { projectMap[p.id] = p; });
      incomingProjects.forEach(p => {
        const existing = projectMap[p.id];
        // If project exists, keep the newer version
        if (!existing || (p.updatedAt || 0) >= (existing.updatedAt || 0)) {
          projectMap[p.id] = p;
        }
      });

      const mergedProjects = Object.values(projectMap);

      // Merge manufacturers, reps, fixtureTypes (union)
      const mergeArrays = (a, b) => [...new Set([...(a||[]), ...(b||[])])];
      const mergeLibrary = (a, b) => {
        const map = {};
        [...(a||[]), ...(b||[])].forEach(f => { map[f.libraryId || f.id] = f; });
        return Object.values(map);
      };

      const merged = {
        projects: mergedProjects,
        manufacturers: mergeArrays(existing.manufacturers, incoming.manufacturers),
        reps: mergeArrays(existing.reps, incoming.reps),
        fixtureTypes: mergeArrays(existing.fixtureTypes, incoming.fixtureTypes),
        library: mergeLibrary(existing.library, incoming.library),
        brand: incoming.brand || existing.brand,
        activeProjectId: incoming.activeProjectId || existing.activeProjectId,
      };

      // Save merged data
      const uploadRes = await fetch(`https://blob.vercel-storage.com/${FILENAME}`, {
        method: "PUT",
        headers: {
          "Authorization": `Bearer ${token}`,
          "Content-Type": "application/json",
          "x-content-type": "application/json",
          "x-add-random-suffix": "0",
        },
        body: JSON.stringify(merged),
      });

      if (!uploadRes.ok) {
        const err = await uploadRes.text();
        return res.status(500).json({ error: err });
      }

      const result = await uploadRes.json();
      return res.status(200).json({ ok: true, url: result.url });
    }

    return res.status(405).json({ error: "Method not allowed" });
  } catch(e) {
    return res.status(500).json({ error: e.message });
  }
};

File: api/sync.py
import json
import os
import urllib.request
import urllib.error
from http.server import BaseHTTPRequestHandler

BLOB_TOKEN = os.environ.get("BLOB_READ_WRITE_TOKEN", "")
STORE_ID = os.environ.get("BLOB_STORE_ID", "")
DATA_FILENAME = "lighting-schedule-data.json"

def get_blob_url():
    # Vercel Blob public URL format
    return f"https://blob.vercel-storage.com"

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        try:
            if not BLOB_TOKEN:
                self._respond(200, {"projects": [], "error": "No blob token"})
                return

            # Try to fetch existing data
            req = urllib.request.Request(
                f"https://blob.vercel-storage.com/{DATA_FILENAME}",
                headers={
                    "Authorization": f"Bearer {BLOB_TOKEN}",
                }
            )
            with urllib.request.urlopen(req, timeout=10) as res:
                raw = res.read().decode("utf-8")
                data = json.loads(raw)
            self._respond(200, data)

        except urllib.error.HTTPError as e:
            if e.code == 404:
                self._respond(200, {"projects": [], "manufacturers": [], "reps": [], "fixtureTypes": [], "library": []})
            else:
                body = e.read().decode() if hasattr(e, 'read') else str(e)
                self._respond(200, {"error": f"HTTP {e.code}: {body}"})
        except Exception as e:
            import traceback
            self._respond(200, {"error": str(e), "trace": traceback.format_exc()})

    def do_POST(self):
        try:
            if not BLOB_TOKEN:
                self._respond(500, {"error": "No blob token configured"})
                return

            length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(length)

            # Upload to Vercel Blob
            req = urllib.request.Request(
                f"https://blob.vercel-storage.com/{DATA_FILENAME}",
                data=body,
                method="PUT",
                headers={
                    "Authorization": f"Bearer {BLOB_TOKEN}",
                    "Content-Type": "application/json",
                    "x-content-type": "application/json",
                    "x-add-random-suffix": "0",
                }
            )
            with urllib.request.urlopen(req, timeout=10) as res:
                result = json.loads(res.read().decode())

            self._respond(200, {"ok": True, "url": result.get("url", "")})

        except Exception as e:
            import traceback
            self._respond(500, {"error": str(e), "trace": traceback.format_exc()})

    def do_OPTIONS(self):
        self._respond(200, {})

    def _respond(self, status, data):
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())

    def log_message(self, format, *args):
        pass
